src/ipld/frozen_cids.rs

```diff
@@ -0,0 +1,204 @@
+// CIDs that would normally be stored as a vector of CIDs. The V1 DAG-CBOR
+// Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid`
+// type) is +99.99% of all CIDs, so very few CIDs need to be stored in the
+// `Heap(Box<Cid>)` variant of `SmallCid`. The Box type has been chosen to make
+// `FrozenCids` explicitly immutable due to the fact that the vectors are parsed
+// from CBOR and are immutable.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FrozenCids(Box<[SmallCid]>);
+
+impl Default for FrozenCids {
+    fn default() -> Self {
+        FrozenCids(Box::new([]))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+enum SmallCid {
+    Heap(Box<Cid>),
+    Inline([u8; 32]),
+}
+
+pub struct FrozenCidsIterator<'a> {
+    buffer: &'a FrozenCids,
+    current_ix: usize,
+}
+
+impl<'a> IntoIterator for &'a FrozenCids {
+    type Item = Cid;
+    type IntoIter = FrozenCidsIterator<'a>;
+    fn into_iter(self) -> Self::IntoIter {
+        FrozenCidsIterator {
+            buffer: self,
+            current_ix: 0,
+        }
+    }
+}
+
+impl Iterator for FrozenCidsIterator<'_> {
+    type Item = Cid;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_ix < self.buffer.0.len() {
+            let cid = &self.buffer.0[self.current_ix];
+            self.current_ix += 1;
+            match cid {
+                SmallCid::Heap(cid) => Some(*cid.clone()),
+                SmallCid::Inline(bytes) => {
+                    let mut cid = [0; BLAKE2B256_SIZE];
+                    cid.copy_from_slice(bytes);
+                    Some(Cid::new_v1(
+                        DAG_CBOR,
+                        multihash::Multihash::wrap(Blake2b256.into(), &cid)
+                            .expect("failed to convert Blake2b digest to V1 DAG-CBOR Blake2b CID"),
+                    ))
+                }
+            }
+        } else {
+            None
+        }
+    }
+}
+
+impl Serialize for FrozenCids {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        Vec::<Cid>::from(self).serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for FrozenCids {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Self::from(<Vec<Cid>>::deserialize(deserializer)?))
+    }
+}
+
+impl FromIterator<Cid> for FrozenCids {
+    fn from_iter<T: IntoIterator<Item = Cid>>(iter: T) -> Self {
+        let mut vec = Vec::new();
+        for i in iter {
+            vec.push(i);
+        }
+        FrozenCids::from(vec)
+    }
+}
+
+impl From<Vec<Cid>> for FrozenCids {
+    fn from(cids: Vec<Cid>) -> Self {
+        let mut small_cids = Vec::with_capacity(cids.len());
+        for cid in cids {
+            match cid.try_into() {
+                Ok(CidVariant::V1DagCborBlake2b(bytes)) => small_cids.push(SmallCid::Inline(bytes)),
```

Comment:
`CidVariant` and `SmallCid` are very similar - can they be one type?

Reply:
Changed `SmallCid` to `CidVariant` and modified `CidVariant` with generic variant.